Add support for combining RequestFilters and ResponseFilters

droustchev opened the pull request:
Multiple filters can now be combined, as long as they extend the respective `CombinableRequestFilter` or `CombinableResponseFilter` trait.

```scala
object ExampleRequestFilterOne extends RequestFilter[String] with CombinableRequestFilter[String] {
  def apply(request: HttpRequest, data: String): RequestFilterOutput = {
    Future.successful(Right(request))
  }
}
 object ExampleRequestFilterTwo extends RequestFilter[Any] {
  def apply(request: HttpRequest, data: Any): RequestFilterOutput = {
    Future.successful(Right(request))
  }
}

val ExampleCombinedRequestFilter = ExampleRequestFilterOne.combine(ExampleRequestFilterTwo)
```

droustchev-pd committed:
Add CombinableRequestFilter

droustchev-pd committed:
Add CombinableResponseFilter

droustchev-pd committed:
Add CombinableRequestFilter to example

dvgica commented on arrivals/src/main/scala/com/pagerduty/arrivals/impl/filter/CombinableRequestFilter.scala:
```diff
@@ -0,0 +1,21 @@
+
+trait CombinableRequestFilter[-RequestData] extends api.filter.RequestFilter[RequestData] { base =>
+
+  def combine[U <: RequestData](filter: RequestFilter[U])(implicit ec: ExecutionContext): RequestFilter[U] = {
```
`T` instead of `U` is a bit more common for type params which don't have any particular meaning.

droustchev replied:
👍

dvgica commented on arrivals/src/main/scala/com/pagerduty/arrivals/impl/filter/CombinableRequestFilter.scala:
```diff
@@ -0,0 +1,21 @@
+
+import scala.concurrent.{ExecutionContext, Future}
+
+trait CombinableRequestFilter[-RequestData] extends api.filter.RequestFilter[RequestData] { base =>
```
I think I suggested `Combinable` but maybe `Composable` is better? It has specific meaning in programming, which is relevant here.

droustchev replied:
👍

dvgica commented on arrivals/src/main/scala/com/pagerduty/arrivals/impl/filter/CombinableRequestFilter.scala:
```diff
@@ -0,0 +1,21 @@
+  def combine[U <: RequestData](filter: RequestFilter[U])(implicit ec: ExecutionContext): RequestFilter[U] = {
+    new RequestFilter[U] {
+      override def apply(request: HttpRequest, data: U): RequestFilterOutput = {
+        base.apply(request, data) flatMap {
```
No-dot "infix" notation is generally discouraged for non-symbolic methods. See https://docs.scala-lang.org/style/method-invocation.html.
```suggestion
        base.apply(request, data).flatMap {
```

droustchev replied:
Good to know, thanks!

dvgica commented on arrivals/src/main/scala/com/pagerduty/arrivals/impl/filter/CombinableResponseFilter.scala:
```diff
@@ -0,0 +1,20 @@
+  def combine[U <: RequestData](filter: ResponseFilter[U])(implicit ec: ExecutionContext): ResponseFilter[U] = {
+    new ResponseFilter[U] {
+      override def apply(request: HttpRequest, response: HttpResponse, data: U): ResponseFilterOutput = {
+        base.apply(request, response, data) flatMap { resp =>
```
```suggestion
        base.apply(request, response, data).flatMap { resp =>
```

dvgica commented on arrivals/src/main/scala/com/pagerduty/arrivals/impl/filter/CombinableResponseFilter.scala:
```diff
@@ -0,0 +1,20 @@
+
+trait CombinableResponseFilter[-RequestData] extends api.filter.ResponseFilter[RequestData] { base =>
+
+  def combine[U <: RequestData](filter: ResponseFilter[U])(implicit ec: ExecutionContext): ResponseFilter[U] = {
```
If people might build long chains of these things, it might be nicer to type fewer characters. I actually think is a good case for a symbolic method. Which one is another question.... to me an arrow makes sense like `->`, but that already has meaning for tuples. `=>` is reserved. `+` is OK but kind of loses the ordering connotation (which is important - this operation isn't guaranteed to be commutative). `~` is already used to compose routes in Akka.

Maybe `~>`? Or if we want something non-symbolic what about `then`? Example:

``` scala
val f = RateLimitFilter ~> AddHeaders ~> Metrics ~> Whatever
```

This is the kind of scenario where symbolic infix operators cut down on the noise.

dvgica commented on arrivals/src/test/scala/com/pagerduty/arrivals/impl/filter/CombinableRequestFilterSpec.scala:
```diff
@@ -0,0 +1,68 @@
+          req.headers.find(_.is(firstHeaderName.toLowerCase)) shouldEqual Some(RawHeader(firstHeaderName, "test"))
+          req.headers.find(_.is(secondHeaderName.toLowerCase)) shouldEqual Some(RawHeader(secondHeaderName, "test"))
+        }
+        case Left(_) => fail("Whoopsie!")
```
Maybe a more specific failure message is in order?

droustchev replied:
I wonder how useful the message is to begin with, since this is just making sure that the test fails if we receive a `HttpResponse` instead of a `HttpRequest` from the first filter. See the output of both a test with a custom message and one without. Again, this test is essentially set up wrong, not actually failing.

```shell
[info] - can be composed with another RequestFilter of the same RequestData type *** FAILED ***
[info]   Received HttpResponse instead of HttpRequest from first request filter (ComposableRequestFilterSpec.scala:44)
[info] - can be composed with another RequestFilter of different RequestData type *** FAILED ***
[info]   org.scalatest.exceptions.TestFailedException was thrown. (ComposableRequestFilterSpec.scala:66)
```

dvgica commented on arrivals/src/test/scala/com/pagerduty/arrivals/impl/filter/CombinableRequestFilterSpec.scala:
```diff
@@ -0,0 +1,68 @@
+      case Left(_) => fail("Whoopsie!")
+    }
+  }
+}
```
What about the case where the first filter returns a `Left(response)`? Seems like we should test that the second filter isn't run.

droustchev replied:
Definitely! I totally forgot about that one.

dvgica commented on arrivals-example/src/main/scala/com/pagerduty/arrivals/example/ExampleGateway.scala:
```diff
@@ -23,8 +25,11 @@ class ExampleGateway(
+
   lazy val routes = pathPrefix("api")(
     prefixProxyRoute("cats", CatsUpstream, ExampleResponseFilter) ~
+      prefixProxyRoute("birds", BirdsUpstream, ExampleCombinedRequestFilter) ~
```
I think it's valuable to keep the example as simple as possible. We already have a route that uses `ExampleResponseFilter` - why not add another filter on to that one? I don't see a need for an additional upstream, additional dummy server, etc.

droustchev replied:
Part of my motivation was to have individual aspects separated and not all cobbled together. I agree that maybe the the added upstream/dummy, etc is maybe counterproductive in that regard.

dvgica reviewed (changes requested):
Looking good, a few comments.

droustchev-pd committed:
Rename type parameter or filter composition method

droustchev-pd committed:
Rename Combinable*Filter to Composable*Filter

droustchev-pd committed:
Do not use no-dot infix notation for flatMap in composable filters

droustchev-pd committed:
Use symbolic method ~> for filter composition

droustchev-pd committed:
Add test for ComposableRequestFilter

If the first filter doesn't return the expected HttpRequest, but a
HttpResponse, we want to make sure we don't apply subsequent filters

droustchev-pd committed:
Revert "Add CombinableRequestFilter to example"

This reverts commit ae98b05c6ed3125efb453b0fcc1dcb15b616e08c.

droustchev-pd committed:
Add example for composing filters

dvgica reviewed (approved):
Nice work!

droustchev-pd committed:
Remove filter composition TODO from README.md

droustchev-pd committed:
Bump version to 0.29.0
